bin/monitor.ml

```diff
@@ -0,0 +1,236 @@
+      state.progress <- progress;
+      State.maybe_reset_diagnostics state;
+      loop ()
+    (* Update the diagnostics. If diagnostics are added then print them. If they are
+       removed then udpate the state and rerender entirely. *)
+    | Event.Diagnostics diagnostics ->
+      List.iter diagnostics ~f:(function
+        | Diagnostic.Event.Add d ->
+          state.diagnostics <- Diagnostic.Id.Map.add_exn state.diagnostics d.id d;
+          Console.print_user_message (Diagnostic.to_user_message d);
+          State.status state (* we make sure to update the status here *)
+        | Remove d ->
+          state.diagnostics <- Diagnostic.Id.Map.remove state.diagnostics d.id;
+          Console.reset ();
+          State.render_state state);
+      loop ()
+    (* If one of the writer loops disconnected, then we must have disconnected from the
+       server. Therefore we tell the user that we will be exiting and loop once more so
+       that we can clear the Mvar. *)
+    | Poller_disconnected ->
+      state.quit <- true;
+      Console.print_user_message
+        (User_error.make [ Pp.textf "Lost connection to server, exiting..." ]);
+      loop ()
+  in
+  loop ()
+;;
+
+(** Main method *)
+
+let exec () =
+  let where = Rpc_common.active_server () in
+  let* connect = Client.Connection.connect_exn where in
+  Dune_rpc_impl.Client.client
+    connect
+    (Dune_rpc_private.Initialize.Request.create
+       ~id:(Dune_rpc_private.Id.make (Sexp.Atom "monitor_cmd")))
+    ~f:(fun client ->
+      let* progress_poller = poll_exn client Dune_rpc_private.Public.Sub.progress
+      and* jobs_poller = poll_exn client Dune_rpc_private.Public.Sub.running_jobs
+      and* diagnostics_poller = poll_exn client Dune_rpc_private.Public.Sub.diagnostic in
+      (* The [event] mvar is shared between the loops. The [render_loop]
+         actively reads from the mvar to update the display and the other loops
+         write to it. *)
+      let (event : Event.t Fiber.Mvar.t) = Fiber.Mvar.create () in
+      Fiber.all_concurrently_unit
+        [ render_loop event
+        ; fetch_loop event jobs_poller ~f:Event.jobs
+        ; fetch_loop event progress_poller ~f:Event.progress
+        ; fetch_loop event diagnostics_poller ~f:Event.diagnostics
+        ])
+;;
+
+let man =
+  [ `S "DESCRIPTION"
+  ; `P
+      {|$(b,dune monitor) connects to an RPC server running in the current
+      workspace and displays the build progress and diagnostics.|}
+  ]
+;;
+
+let command =
+  let info =
+    let doc = "Connect to a Dune RPC server and monitor it." in
+    Cmd.info "monitor" ~doc ~man
+  and term =
+    let open Import in
+    let+ (common : Common.t) = Common.term in
+    let common = Common.forbid_builds common in
+    let config = Common.init ~log_file:No_log_file common in
+    let stats = Common.stats common in
+    let config =
+      let insignificant_changes = Common.insignificant_changes common in
+      let signal_watcher = `Yes in
```

Comment:
Well I think I need it because I have no way of sending `Ctrl + C` when in the monitor. When I have it, that can be used to quit the monitor. Without it I get stuck.

Reply:
Something is seriously broken then. Initializing the signal watcher is what captures the `Ctrl + C`. By default, it should just be terminating the program.